src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,421 @@
+/-
+Copyright (c) 2020 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import group_theory.congruence
+import linear_algebra.multilinear
+
+/-!
+# Tensor product of an indexed family of semimodules over commutative semirings
+
+We define the tensor product of an indexed family `s : ι → Type*` of semimodules over commutative
+semirings. We denote this space by `⨂[R] i, s i` and define it as `free_add_monoid (R × Π i, s i)`
+quotiented by the appropriate equivalence relation. The treatment follows very closely that of the
+binary tensor product in `linear_algebra/tensor_product.lean`.
+
+## Main definitions
+
+* `pi_tensor_product R s` with `R` a commutative semiring and `s : ι → Type*` is the tensor product
+  of all the `s i`'s. This is denoted by `⨂[R] i, s i`.
+* `tprod R f` with `f : Π i, s i` is the tensor product of the vectors `f i` over all `i : ι`.
+* `mk R s` is the canonical multilinear map from `Π i, s i` to `⨂[R] i, s i`.
+* `lift_add_hom` constructs an `add_monoid_hom` from `(⨂[R] i, s i)` to some space `F` from a
+function `φ : (R × Π i, s i) → F` with the appropriate properties.
+* `lift φ` with `φ : multilinear_map R s E` is the corresponding linear map
+  `(⨂[R] i, s i) →ₗ[R] E`.
+
+## Notations
+
+* `⨂[R] i, s i` is defined as localized notation in locale `tensor_product`
+* `⨂ₜ[R] i, f i` with `f : Π i, f i` is defined globally as the tensor product of all the `f i`'s.
+
+## Implementation notes
+
+* We define it via `free_add_monoid (R × Π i, s i)` with the `R` representing a "hidden" tensor
+factor, rather than `free_add_monoid (R × Π i, s i)` to ensure that, if `ι` is an empty type,
+the space is isomorphic to the base ring `R`.
+* We have not restricted the index type `ι` to be a `fintype`, as nothing we do here strictly
+requires it. However, problems may arise in the case where `ι` is infinite; use at your own caution.
```

Comment:
```suggestion
* We define it via `free_add_monoid (R × Π i, s i)` with the `R` representing a "hidden" tensor
  factor, rather than `free_add_monoid (R × Π i, s i)` to ensure that, if `ι` is an empty type,
  the space is isomorphic to the base ring `R`.
* We have not restricted the index type `ι` to be a `fintype`, as nothing we do here strictly
  requires it. However, problems may arise in the case where `ι` is infinite; use at your own caution.
```